README.md
# clj-foundationdb [![Build Status](https://travis-ci.org/tirkarthi/clj-foundationdb.svg?branch=master)](https://travis-ci.org/tirkarthi/clj-foundationdb) [![Clojars Project](https://img.shields.io/clojars/v/xtreak/clj-foundationdb.svg)](https://clojars.org/xtreak/clj-foundationdb) [![cljdoc](https://cljdoc.xyz/badge/xtreak/clj-foundationdb)](https://cljdoc.xyz/d/xtreak/clj-foundationdb)

A Clojure wrapper for FoundationDB

## Documentation

cljdoc version : https://cljdoc.xyz/d/xtreak/clj-foundationdb/

codox version : https://tirkarthi.github.io/clj-foundationdb


I am also trying to port a class scheduling app to Clojure using the library. Currently the API is still in a lot of fluctuation and hence the examples might not work but you can view the source at examples folder.

Reference implementation in Java : https://apple.github.io/foundationdb/class-scheduling-java.html

## Examples

```clojure

(ns examples.main
  (:require [clj-foundationdb.core :refer :all]
            [clj-foundationdb.utils :refer :all]))

;; Set a key

(let [fd    (select-api-version 520)
      key   "foo"
      value 1]
  (with-open [db (open fd)]
    (tr! db
         (set-val tr key value))))

nil

;; Get a key

(let [fd    (select-api-version 520)
      key   "foo"]
  (with-open [db (open fd)]
    (tr! db
         (get-val tr key))))

1

;; Perform multiple operations in a single transaction

(let [fd    (select-api-version 520)
      key   "foo"
      value 1]
  (with-open [db (open fd)]
    (tr! db
         (set-val tr key value)
         (get-val tr key))))

1

;; Set multiple keys with same value

(let [fd    (select-api-version 520)
      key   [["bar"] ["car"] ["dar"] ["far"]]
      value 1]
  (with-open [db (open fd)]
    (tr! db
         (set-keys tr key value))))

nil

;; Get a range of keys

(let [fd    (select-api-version 520)
      begin "car"
      end   "far"]
  (with-open [db (open fd)]
    (tr! db
         (get-range tr begin end))))

[[["car"] 1] [["dar"] 1]]

;; Get all keys

(let [fd    (select-api-version 520)]
  (with-open [db (open fd)]
    (tr! db
         (get-all tr))))

[[["bar"] 1] [["car"] 1] [["dar"] 1] [["far"] 1] [["foo"] 1]]

;; First key less than given key

(let [fd    (select-api-version 520)
      key   "car"]
  (with-open [db (open fd)]
    (tr! db
         (last-less-than tr key))))

[[["bar"] 1]]

;; First key greater than given key

(let [fd    (select-api-version 520)
      key   "car"]
  (with-open [db (open fd)]
    (tr! db
         (first-greater-than tr key))))

[[["dar"] 1]]

;; Nested keys

(let [fd      (select-api-version 520)
      classes [["class" "intro"] ["class" "algebra"] ["class" "maths"] ["class" "bio"]]
      time    "10:00"]
  (with-open [db (open fd)]
    (tr! db
         (set-keys tr classes time)
         (get-val tr ["class" "algebra"]))))

"10:00"

;; Automatic subspace prefix within a context

(let [fd      (select-api-version 520)
      classes ["intro" "algebra" "maths" "bio"]
      time    "10:00"]
  (with-open [db (open fd)]
    (tr! db
         (with-subspace "class"
           (mapv #(set-val tr %1 time) classes)
           (get-val tr "algebra")))))

"10:00"
```

## Transaction basic example

A basic example of the transactional nature of FoundationDB can be explained as below :

* Set a key "foo" with value "1" and inside the transaction the value is set as "1"
* Try changing the value to "2". You can see that inside the transaction the value of "foo" is "2" once it's changed. But when there is an exception the value is rolled back to "1"
* Get the value of "foo" and it should return "1" since the previous transaction aborted with an exception and hence the value is not updated

```clojure
clj-foundationdb.core> (let [fd    (select-api-version 520)
                             key   "foo"
                             value "1"]
                         (with-open [db (open fd)]
                           (tr! db
                                (set-val tr key value)
                                (get-val tr key))))

1

clj-foundationdb.core> (let [fd    (select-api-version 520)
                             key   "foo"
                             value "2"]
                         (with-open [db (open fd)]
                           (tr! db
                                (println (get-val tr key))
                                (set-val tr key value)
                                (println (get-val tr key))
                                (/ 1 0))))

1
2
ArithmeticException Divide by zero  clojure.lang.Numbers.divide (Numbers.java:163)

clj-foundationdb.core> (let [fd    (select-api-version 520)
                             key   "foo"
                             value "1"]
                         (with-open [db (open fd)]
                           (tr! db
                                (get-val tr key))))
1
```

## Tuple encoding and alternatives

This library uses tuple encoding so that manual encoding and decoding is not necessary. If you want a more thin wrapper please use [clj_fdb](https://github.com/vedang/clj_fdb/). Relevant issue : https://github.com/tirkarthi/clj-foundationdb/issues/15

## Stability

The API is still in early development stages and is subjected to change. It's building up nicely and comments are welcome.

### JDK 7 support

The underlying Java library uses lambdas and hence the minimum supported version is JDK 8.

## Benchmarks

Benchmarks are quite tricky to get right. The main factor with FoundationDB is that everything opens up a new transaction and hence making it slow. But bulk writes are faster.

### System Info

```
➜  clj-foundationdb git:(master) ✗ system_profiler SPHardwareDataType
Hardware:

    Hardware Overview:

      Model Name: MacBook Air
      Model Identifier: MacBookAir7,2
      Processor Name: Intel Core i5
      Processor Speed: 1.6 GHz
      Number of Processors: 1
      Total Number of Cores: 2
      L2 Cache (per Core): 256 KB
      L3 Cache: 3 MB
      Memory: 4 GB
      Boot ROM Version: MBA71.0166.B06
      SMC Version (system): 2.27f2
      Serial Number (system): C02Q4LHTG940
      Hardware UUID: 4ECF3B7B-6444-51AA-BB60-BCE6C285D90D
```

### FoundationDB single writes

```
(let [fd    (select-api-version 520)
      key   "foo"
      value "1"]
      (with-open [db (open fd)]
	        (quick-bench (doall (for [_ (range 100)] (tr! db (set-val tr key value)))))))
Evaluation count : 6 in 6 samples of 1 calls.
             Execution time mean : 1.845528 sec
    Execution time std-deviation : 58.156807 ms
   Execution time lower quantile : 1.795201 sec ( 2.5%)
   Execution time upper quantile : 1.934851 sec (97.5%)
                   Overhead used : 2.028213 ns

Found 1 outliers in 6 samples (16.6667 %)
	low-severe	 1 (16.6667 %)
 Variance from outliers : 13.8889 % Variance is moderately inflated by outliers
```

### FoundationDB batch writes


```
(defn set-val-n
  "Set a value for the key"
  [tr key value n]
  (dotimes [_ n]
     (.set tr key value)))

clj-foundationdb.core> (let [fd (select-api-version 520)
                             key (.getBytes "foo")
                             value (.getBytes "10000")]
                         (with-open [db (open fd)]
                           (quick-bench (tr! db (set-val-n tr key value 100000)))))
Evaluation count : 6 in 6 samples of 1 calls.
             Execution time mean : 625.745197 ms
    Execution time std-deviation : 72.918243 ms
   Execution time lower quantile : 542.717590 ms ( 2.5%)
   Execution time upper quantile : 718.452721 ms (97.5%)
                   Overhead used : 2.028213 ns

```

### FoundationDB write 1 million keys with 10 parallel clients and 100k keys per client

1 million keys set in 23.74 seconds

```clojure
(let [fd (select-api-version 520)
      kv (map #(vector (str %1) %1) (range 100000))]
  (time (let [clients (repeatedly 10 #(future
                                        (with-open [db (open fd)]
                                          (tr! db
                                               (doall (doseq [[k v] kv]
                                                        (set-val tr k v)))))))]
          (doall (map deref clients))
          "Finished")))
"Elapsed time: 23740.242199 msecs"
"Finished"
```

### FoundationDB read 100k keys with 10 parallel clients and 10k keys per client

100k keys read took 16.81 seconds

```
(let [fd (select-api-version 520)
      kv (map #(vector (str %1) %1) (range 100000))]
  (time (with-open [db (open fd)]
          (let [clients (repeatedly 10 #(future
                                          (tr! db
                                               (doall (doseq [k (range 10000)]
                                                        (get-val tr k))))))]
            (doall (map deref clients)))
          "Finished")))
"Elapsed time: 16818.560737 msecs"
"Finished"
```

You can find some more numbers for FoundationDB [here](https://apple.github.io/foundationdb/benchmarking.html)

### Redis

```
➜  redis git:(unstable) redis-benchmark -t set
====== SET ======
  100000 requests completed in 1.59 seconds
  50 parallel clients
  3 bytes payload
  keep alive: 1

96.52% <= 1 milliseconds
99.95% <= 2 milliseconds
99.97% <= 6 milliseconds
100.00% <= 6 milliseconds
62853.55 requests per second
```

### Redis with pipelining

```
➜  redis git:(unstable) redis-benchmark -t set -P 160
====== SET ======
  100000 requests completed in 0.17 seconds
  50 parallel clients
  3 bytes payload
  keep alive: 1

0.00% <= 1 milliseconds
0.64% <= 2 milliseconds
1.12% <= 5 milliseconds
2.72% <= 6 milliseconds
4.48% <= 10 milliseconds
8.48% <= 11 milliseconds
36.96% <= 12 milliseconds
65.12% <= 13 milliseconds
82.40% <= 14 milliseconds
95.36% <= 15 milliseconds
99.04% <= 16 milliseconds
100.00% <= 16 milliseconds
595238.12 requests per second
```

## License

Copyright © 2018 Karthikeyan S

Distributed under the MIT License


CHANGELOG.md
# Change Log

## [0.0.1]

- Add watch, set-tuple-vals, get-tuple-range-startswith, first-greater-than, first-greater-or-equal, last-less-than, last-less-or-equal
- Add clear-key, get-range, get-range-startswith, get-all, clear-range
- Add method to set a value, get a value, get a range of keys, clear database


test/clj_foundationdb/core_test.clj
(ns clj-foundationdb.core-test
  (:import (com.apple.foundationdb Database)
           (com.apple.foundationdb FDB)
           (com.apple.foundationdb Range)
           (com.apple.foundationdb.tuple Tuple)
           (java.nio ByteBuffer))
  (:require [clojure.test :refer :all]
            [clj-foundationdb.core :refer :all]
            [clj-foundationdb.utils :refer :all]))

;; Somehow I am not able to bind the db connection so that it's accessible in the test resulting
;; in a transaction for write and another transaction for test in the testing function. Then
;; finally one cleaning up the fixture.
;; Reusing the db to open new transaction also results in the fixture data not being committed
;; and hence not being read by the testing function transaction

(defn key-fixture
  [test]
  (let [fd    (select-api-version 520)
        keys  ["bar" "bar1" "bar2" "car"]
        value "1"]
    (with-open [db (open fd)]
      (tr! db
           (set-keys tr keys value)))
    (test)
    (with-open [db (open fd)]
      (tr! db
           (clear-all tr)))))

(use-fixtures :each key-fixture)

(deftest test-transaction
  (testing "Test transaction rollback"
    (let [fd    (select-api-version 520)
          key   "foo"
          value "1"]
      (with-open [db (open fd)]
        (tr! db
             (is (nil? (get-val tr key)))
             (set-val tr key value)
             (is (= (get-val tr key) value)))))
    (let [fd    (select-api-version 520)
          key   "foo"
          value "2"]
      (with-open [db (open fd)]
        (is (thrown? Exception
                     (tr! db
                          (set-val tr key value)
                          (is (= (get-val tr key) value))
                          (/ 1 0))))))
    (let [fd    (select-api-version 520)
          key   "foo"
          value "1"]
      (with-open [db (open fd)]
        (tr! db
             (is (= (get-val tr key) value)))))))

(deftest test-set
  (testing "Test simple set with string"
    (let [fd    (select-api-version 520)
          key   "foo"
          value "1"]
      (with-open [db (open fd)]
        (tr! db
             (is (nil? (get-val tr key)))
             (set-val tr key value)
             (is (= (get-val tr key) value))))))

  (testing "Test simple set with int"
    (let [fd    (select-api-version 520)
          key   "foo"
          value 1]
      (with-open [db (open fd)]
        (tr! db
             (set-val tr key value)
             (is (= (get-val tr key) value))))))

  (testing "Test simple set with float"
    (let [fd    (select-api-version 520)
          key   "foo"
          value 1.5]
      (with-open [db (open fd)]
        (tr! db
             (set-val tr key value)
             (is (= (get-val tr key) value))))))

  (testing "Test simple set with vector"
    (let [fd    (select-api-version 520)
          key   "foo"
          value [1 2 3]]
      (with-open [db (open fd)]
        (tr! db
             (set-val tr key value)
             (is (= (get-val tr key) 1))
             (is (= (get-val tr key :coll true) value)))))))

(deftest test-multiple-set
  (testing "Test multiple set"
    (let [fd            (select-api-version 520)
          keys          ["foo" "for"]
          expected-keys (seq ["bar" "bar1" "bar2" "car" "foo" "for"])
          value         "1"]
      (with-open [db (open fd)]
        (tr! db
             (set-keys tr keys value)
             (is (every? #(= (get-val tr %1) value) expected-keys))))))

  (testing "Test get all keys"
    (let [fd    (select-api-version 520)
          keys  (seq ["bar" "bar1" "bar2" "car" "foo" "for"])
          value "1"]
      (with-open [db (open fd)]
        (tr! db
             (let [returned-keys (mapcat first (get-all tr))]
               (is (= keys returned-keys))
               (is (every? #(= (get-val tr %1) value) keys))))))))

(deftest test-get-range-begin-end
  (testing "Test get all keys with the range of begin and end with end being exclusive. [a, b)"
    (let [fd     (select-api-version 520)
          prefix "b"
          value  "1"]
      (with-open [db (open fd)]
        (tr! db
             (let [keys (mapcat first (get-range tr "b" "bar2"))]
               (is (= '("bar" "bar1") keys))
               (is (every? #(= (get-val tr %1) value) keys))))))))

(deftest test-get-range-starts-with
  (testing "Test get all keys with prefix"
    (let [fd     (select-api-version 520)
          prefix "car"
          value  "1"]
      (with-open [db (open fd)]
        (tr! db
             (let [keys (mapcat first (get-range-startswith tr prefix))]
               (is (= '("car") keys))
               (is (every? #(= (get-val tr %1) value) keys))))))))

(deftest test-get-last-less-than
  (testing "Test last less than the given key"
    (let [fd            (select-api-version 520)
          key           "bar1"
          value         "1"
          expected-keys '("bar")]
      (with-open [db (open fd)]
        (tr! db
             (let [keys (mapcat first (last-less-than tr key))]
               (is (= expected-keys keys))
               (is (every? #(= (get-val tr %1) value) keys))))))))

(deftest test-get-last-less-or-equal
  (testing "Test last less than or equal to the given key"
    (let [fd            (select-api-version 520)
          key           "bar1"
          value         "1"
          expected-keys '("bar1")]
      (with-open [db (open fd)]
        (tr! db
             (let [keys (mapcat first (last-less-or-equal tr key))]
               (is (= expected-keys keys))
               (is (every? #(= (get-val tr %1) value) keys))))))))

(deftest test-get-first-greater-than
  (testing "Test first greater than"
    (let [fd            (select-api-version 520)
          key           "bar"
          value         "1"
          expected-keys '("bar1")]
      (with-open [db (open fd)]
        (tr! db
             (let [keys (mapcat first (first-greater-than tr key))]
               (is (= expected-keys keys))
               (is (every? #(= (get-val tr %1) value) keys)))))))

  (testing "Test first greater than with limit"
    (let [fd            (select-api-version 520)
          key           "bar"
          value         "1"
          limit         2
          expected-keys '("bar1" "bar2")]
      (with-open [db (open fd)]
        (tr! db
             (let [keys (mapcat first (first-greater-than tr key limit))]
               (is (= expected-keys keys))
               (is (every? #(= (get-val tr %1) value) keys)))))))

  (testing "Test first greater than with last key"
    (let [fd    (select-api-version 520)
          key   "foo"
          value "1"]
      (with-open [db (open fd)]
        (tr! db
             (let [keys (first-greater-than tr key)]
               (is (empty? keys))))))))

(deftest test-get-first-greater-or-equal
  (testing "Test first greater or equal"
    (let [fd            (select-api-version 520)
          key           "bar"
          value         "1"
          expected-keys '("bar")]
      (with-open [db (open fd)]
        (tr! db
             (let [keys (mapcat first (first-greater-or-equal tr key))]
               (is (= expected-keys keys))
               (is (every? #(= (get-val tr %1) value) keys)))))))

  (testing "Test first greater or equal with limit"
    (let [fd            (select-api-version 520)
          key           "bar"
          value         "1"
          limit         2
          expected-keys '("bar" "bar1")]
      (with-open [db (open fd)]
        (tr! db
             (let [keys (mapcat first (first-greater-or-equal tr key limit))]
               (is (= expected-keys keys))
               (is (every? #(= (get-val tr %1) value) keys))))))))

(deftest test-non-existent-get
  (testing "Test non-existent record for nil"
    (let [fd  (select-api-version 520)
          key "foo12"]
      (with-open [db (open fd)]
        (tr! db
             (is (nil? (get-val tr key))))))))

(deftest test-clear
  (testing "Test clear a key"
    (let [fd    (select-api-version 520)
          key   "foo123"
          value "1"]
      (with-open [db (open fd)]
        (tr! db
             (set-val tr key value)
             (is (= (get-val tr key) "1"))
             (clear-key tr key)
             (is (nil? (get-val tr key))))))))

(deftest test-clear-range
  (testing "Test clearing range"
    (let [fd      (select-api-version 520)
          in-keys [["foo" "a"] ["foo" "b"] ["gum" "c"]]
          value   "1"
          k       ["gum" "c"]]
      (with-open [db (open fd)]
        (tr! db
             (set-keys tr in-keys value)
             (is (every? #(= (get-val tr %1) value) in-keys))
             (clear-range tr "foo")
             (is (every? #(nil? (get-val tr %1)) (butlast in-keys)))
             (is (= (get-val tr k) value))))))

  (testing "Test clearing range with begin and end with end being exclusive"
    (let [fd      (select-api-version 520)
          in-keys [["bar" "a"] ["foo" "a"] ["foo" "b"] ["gum" "a"]]
          value   "1"]
      (with-open [db (open fd)]
        (tr! db
             (set-keys tr in-keys value)
             (is (every? #(= (get-val tr %1) value) in-keys))
             (clear-range tr "bar" "gum")
             (is (every? #(nil? (get-val tr %1)) (butlast in-keys))))))))

(deftest test-clear-all
  (testing "Test clear all"
    (let [fd    (select-api-version 520)
          key   "z"
          value "1"]
      (with-open [db (open fd)]
        (tr! db
             (set-val tr key value)
             (clear-all tr)
             (is (empty? (get-all tr))))))))

(deftest test-subspace
  (testing "Test simple subspace"
    (let [fd    (select-api-version 520)
          key   "foo"
          value "bar"]
      (with-open [db (open fd)]
        (tr! db
             (clear-all tr)
             (with-subspace "class"
               (set-val tr key value)
               (is (= (get-val tr key) value)))
             (is (nil? (get-val tr key)))
             (is (= (get-all tr) [[["class" "foo"] "bar"]]))))))

  (testing "Test multiple level subspace"
    (let [fd    (select-api-version 520)
          key   "foo"
          value "bar"]
      (with-open [db (open fd)]
        (tr! db
             (clear-all tr)
             (with-subspace ["class" "intro"]
               (set-val tr key value)
               (is (= (get-val tr key) value)))
             (is (nil? (get-val tr key)))
             (is (= (get-range tr ["class"]) [[["class" "intro" "foo"] "bar"]]))
             (is (= (get-all tr) [[["class" "intro" "foo"] "bar"]])))))))

(deftest test-watch
  (testing "Test watch and callback is called"
    (let [fd    (select-api-version 520)
          key    "foo"
          value  "bar"]
      (with-open [db (open fd)]
        (tr! db
             (clear-key tr key)
             (let [callback (fn [] :called)]
               (future (do (Thread/sleep 100) (set-val tr key value)))
               (is (= :called @(watch tr key callback)))))))))


src/clj_foundationdb/core.clj
(ns clj-foundationdb.core
  (:require [clojure.spec.alpha :as spec]
            [clj-foundationdb.utils :refer :all])
  (:import (com.apple.foundationdb Database
                                   FDB
                                   Range
                                   KeySelector
                                   Transaction)
           (com.apple.foundationdb.tuple Tuple)
           (com.apple.foundationdb.subspace Subspace)
           (java.util List)))

(def tr? #(instance? com.apple.foundationdb.Database %1))

;; Refer https://github.com/apple/foundationdb/blob/e0c8175f3ccad92c582a3e70e9bcae58fff53633/bindings/java/src/main/com/apple/foundationdb/tuple/TupleUtil.java#L171

(def serializable? #(or (number? %1)
                        (string? %1)
                        (decimal? %1)
                        (instance? List %1)))

(defmacro tr!
  "Transaction macro to perform actions. Always use tr for actions inside
  each action since the transaction variable is bound to tr in the functions.

  (let [fd    (select-api-version 520)
        key   \"foo\"
        value \"bar\"]
  (with-open [db (open fd)]
     (tr! db
          (set-val tr key value)
          (get-val tr key))))
  "
  [db & actions]
  `(.run ~db
         (reify
           java.util.function.Function
           (apply [this tr]
             ~@actions))))

(def ^:dynamic *subspace* nil)

(defn make-subspace
  "
  Returns a key with name as prefix

  (make-subspace [\"class\" \"intro\"] [\"algebra\"]) returns (\"class\" \"intro\" \"algebra\")
  "
  [prefix key]
  (flatten (map vector [prefix key])))

(defmacro with-subspace
  "
  Sets and gets the keys with the given subspace key prefixed.
  This essentially executes with code binding the given prefix to *subspace*.

  (let [fd    (select-api-version 520)
        key   \"foo\"
        value \"bar\"]
    (with-open [db (open fd)]
      (tr! db
           (clear-all tr)
           (with-subspace \"class\"
             (set-val tr key value)
             (get-val tr key))
            (nil? (get-val tr key)))))
  "
  [prefix & actions]
  `(binding [*subspace* ~prefix]
     ~@actions))

(defn get-val
  "Get the value for the given key. Accepts the below :

  :subspace - Subspace to be prefixed
  :coll     - Boolean to indicate if the value needs to be deserialized as collection

  (let [fd  (select-api-version 520)
        key \"foo\"]
  (with-open [db (open fd)]
     (tr! db
          (get-val tr key))))

  (let [fd    (select-api-version 520)
       key   \"foo\"
       value [1 2 3]]
  (with-open [db (open fd)]
    (tr! db
         (set-val tr key value)
         (get-val tr key) ;; 1
         (get-val tr key :coll true)))) ;; [1 2 3]
  "
  [tr key & {:keys [subspace coll] :or {subspace *subspace* coll false}}]
  (let [key   (-> (if subspace (make-subspace subspace key) key)
                  key->packed-tuple)]
    (if-let [value @(.get tr key)]
      (if coll
        (.getItems (Tuple/fromBytes value))
        (.get (Tuple/fromBytes value) 0)))))

(spec/fdef get-val
           :args (spec/cat :tr tr? :key serializable?)
           :ret (spec/nilable serializable?))

(defn set-val
  "Set a value for the key. Accepts the below :

  :subspace - Subspace to be prefixed

  (let [fd    (select-api-version 520)
        key   \"foo\"
        value \"bar\"]
  (with-open [db (open fd)]
     (tr! db
          (set-val tr key value))))
  "
  [tr key value & {:keys [subspace] :or {subspace *subspace*}}]
  (let [key   (-> (if subspace (make-subspace subspace key) key)
                  key->packed-tuple)
        value (key->packed-tuple value)]
    (.set tr key value)))

(spec/fdef set-val
           :args (spec/cat :tr tr? :key serializable? :value serializable?)
           :ret (spec/nilable serializable?))

(defn set-keys
  "Set given keys with the value

  (let [fd    (select-api-version 520)
        keys  [\"foo\" \"baz\"]
        value \"bar\"]
  (with-open [db (open fd)]
     (tr! db (set-keys tr keys value))))
  "
  [tr keys value]
  (let [keys  (map #(key->packed-tuple %1) keys)
        value (key->packed-tuple value)]
    (doseq [key keys] (.set tr key value))))

(spec/fdef set-keys
           :args (spec/cat :tr tr? :key (spec/coll-of serializable?) :value serializable?))

(defn clear-key
  "Clear a key from the database

  (let [fd  (select-api-version 520)
        key \"foo\"]
  (with-open [db (open fd)]
     (tr! db (clear-key tr key))))
  "
  [tr key]
  (let [key (key->packed-tuple key)]
    (.clear tr key)))

(spec/fdef clear-key
           :args (spec/cat :tr tr? :key serializable?))

(defn get-range-startswith
  "Get a range of key values as a vector that starts with prefix

  (let [fd     (select-api-version 520)
        prefix \"f\"]
  (with-open [db (open fd)]
     (tr! db (get-range-startswith tr key prefix))))
  "
  [tr prefix]
  (let [prefix      (key->packed-tuple prefix)
        range-query (Range/startsWith prefix)]
    (->> (.getRange tr range-query)
         range->kv)))

(spec/fdef get-range-startswith
           :args (spec/cat :tr tr? :prefix serializable?))

(defn watch
  "
  A key to watch and a callback function to be executed on change. It returns a future object
  that is realized when there is a change to key. Change in key value or clearing the key
  is noted as a change. A set statement with the old value is not a change.

  (let [fd    (select-api-version 520)
       key    \"foo\"
       value  \"bar\"]
  (with-open [db (open fd)]
    (tr! db
         (clear-key tr key)
         (watch tr key #(println \"key is set\"))
         (set-val tr key value)
         (watch tr key #(println \"key is changed to 1\"))
         (set-val tr key value) ;; Doesn't trigger watch
         (set-val tr key \"1\")
         (watch tr key #(println \"cleared key\"))
         (clear-key tr key))))

  key is set
  key is changed to 1
  cleared key
  "
  [tr key callback]
  (let [key   (key->packed-tuple key)
        watch (.watch tr key)]
    (future (do (.join watch)
                (callback)))))

(spec/fdef watch
           :args (spec/cat :tr tr? :key serializable? :callback ifn?))

(defn get-range
  "Get a range of key values as a vector

  (let [fd    (select-api-version 520)
        begin \"foo\"
        end   \"foo\"]
  (with-open [db (open fd)]
     (tr! db (get-range tr begin end))))
  "
  ([tr begin]
   (let [begin       (Tuple/from (to-array (if (sequential? begin) begin [begin])))
         range-query (.getRange tr (.range begin))]
     (range->kv range-query)))
  ([tr begin end]
   (let [begin       (key->packed-tuple begin)
         end         (key->packed-tuple end)
         range-query (.getRange tr (Range. begin end))]
     (range->kv range-query))))

(spec/fdef get-range
           :args (spec/cat :tr tr? :begin serializable? :end serializable?))

;; https://stackoverflow.com/a/21421524/2610955
;; Refer : https://forums.foundationdb.org/t/how-to-clear-all-keys-in-foundationdb-using-java/351/2

(defn get-all
  "Get all key values as a vector"
  [tr]
  (let [begin       (byte-array [])
        end         (byte-array [0xFF])
        range-query (.getRange tr (Range. begin end))]
    (range->kv range-query)))

(spec/fdef get-all
           :args (spec/cat :tr tr?))

(defn clear-range
  "Clear a range of keys from the database.
  When only begin is given then the keys with starting with the tuple are cleared.
  When begin and end are specified then end is exclusive of the range to be cleared.

  (let [fd    (select-api-version 520)
        begin \"foo\"]
  (with-open [db (open fd)]
     (tr! db (clear-range tr begin))))

  (let [fd    (select-api-version 520)
        begin \"foo\"
        end   \"foo\"]
  (with-open [db (open fd)]
     (tr! db (clear-range tr begin end))))
  "
  ([tr begin]
   (let [begin (key->tuple begin)]
     (.clear tr (.range begin))))
  ([tr begin end]
   (let [begin (key->packed-tuple begin)
         end   (key->packed-tuple end)]
     (.clear tr (Range. begin end)))))

(spec/fdef clear-range
           :args (spec/cat :tr tr? :begin serializable? :end serializable?))

;; https://stackoverflow.com/a/21421524/2610955
;; Refer : https://forums.foundationdb.org/t/how-to-clear-all-keys-in-foundationdb-using-java/351/2

(defn clear-all
  "Clear all  keys from the database"
  [tr]
  (let [begin (byte-array [])
        end   (byte-array [0xFF])]
    (.clear tr (Range. begin end))))

(spec/fdef clear-all
           :args (spec/cat :tr tr?))

(defn last-less-than
  "Returns key and value pairs with keys less than the given key for the given limit

  (let [fd  (select-api-version 520)
        key \"foo\"]
  (with-open [db (open fd)]
     (tr! db (last-less-than tr key))))
  "
  ([tr key]
   (last-less-than tr key 1))
  ([tr key limit]
   (let [key         (KeySelector/lastLessThan (key->packed-tuple key))
         end         (.add key limit)
         range-query (.getRange tr key end)]
     (range->kv range-query))))

(spec/fdef last-less-than
           :args (spec/cat :tr tr? :key serializable? :limit (spec/? pos-int?))
           :ret (spec/coll-of (spec/tuple serializable? serializable?)))

(defn last-less-or-equal
  "Returns key and value pairs with keys less than or equal the given key for the given limit

  (let [fd  (select-api-version 520)
        key \"foo\"]
  (with-open [db (open fd)]
     (tr! db (last-less-or-equal tr key))))
  "
  ([tr key]
   (last-less-or-equal tr key 1))
  ([tr key limit]
   (let [key         (KeySelector/lastLessOrEqual (key->packed-tuple key))
         end         (.add key limit)
         range-query (.getRange tr key end)]
     (range->kv range-query))))

(spec/fdef last-less-or-equal
           :args (spec/cat :tr tr? :key serializable? :limit (spec/? pos-int?))
           :ret (spec/coll-of (spec/tuple serializable? serializable?)))

(defn first-greater-than
  "Returns key and value pairs with keys greater than the given key for the given limit

  (let [fd  (select-api-version 520)
        key \"foo\"]
  (with-open [db (open fd)]
     (tr! db (first-greater-than tr key))))
  "
  ([tr key]
   (first-greater-than tr key 1))
  ([tr key limit]
   (let [key         (KeySelector/firstGreaterThan (key->packed-tuple key))
         end         (.add key limit)
         range-query (.getRange tr key end)]
     (range->kv range-query))))

(spec/fdef first-greater-than
           :args (spec/cat :tr tr? :key serializable? :limit (spec/? pos-int?))
           :ret (spec/coll-of (spec/tuple serializable? serializable?)))

(defn first-greater-or-equal
  "Returns key and value pairs with keys greater than or equal to the given key for the given limit

  (let [fd  (select-api-version 520)
        key \"foo\"]
  (with-open [db (open fd)]
     (tr! db (first-greater-or-equal tr key))))
  "
  ([tr key]
   (first-greater-or-equal tr key 1))
  ([tr key limit]
   (let [key         (KeySelector/firstGreaterOrEqual (key->packed-tuple key))
         end         (.add key limit)
         range-query (.getRange tr key end)]
     (range->kv range-query))))

(spec/fdef first-greater-or-equal
           :args (spec/cat :tr tr? :key serializable? :limit (spec/? pos-int?))
           :ret (spec/coll-of (spec/tuple serializable? serializable?)))


examples/class-scheduling/main.clj
(ns class-scheduling.main
  (:import (com.apple.foundationdb Database)
           (com.apple.foundationdb FDB)
           (com.apple.foundationdb Range)
           (com.apple.foundationdb.tuple Tuple)
           (java.nio ByteBuffer))
  (:require [clj-foundationdb.core :refer :all]
            [clj-foundationdb.utils :refer :all]))

(def times ["2:00" "3:00" "4:00"
            "5:00" "6:00" "7:00"
            "8:00" "9:00" "10:00"
            "11:00" "12:00" "13:00"
            "14:00" "15:00" "16:00"
            "17:00" "18:00" "19:00"])

(def types ["chem" "bio" "cs"
            "geometry" "calc" "alg"
            "film" "music" "art" "dance"])

(def levels ["intro" "for dummies" "remedial"
             "101" "201" "301" "mastery"
             "lab" "seminar"])

(def all-classes
  (for [level levels
        type  types
        time  times]
    (clojure.string/join " " [level type time])))

(defn init-db
  [db]
  (tr! db
       (clear-all tr)))

(defn add-class
  [db classes]
  (tr! db
       (set-keys tr (map #(vector "class" %1) classes) 100)))

(defn current-class-attendance
  [db class-name]
  (let [record  ["class" class-name]]
    (tr! db
         (get-val tr record))))

(defn signup-class
  [db student class-name]
  (let [student-record ["attends" student]
        record        ["attends" student class-name]
        seats         (current-class-attendance db class-name)
        class-record  ["class" class-name]
        flag          ""]
    (tr! db
         (if (> (count (get-range tr student-record)) 5)
           (println "Student limit exceeded")
           (do
             (if (and (nil? (get-val tr record))
                      (not (zero? seats)))
               (do
                 (set-val tr record flag)
                 (set-val tr class-record (dec (int seats)))
                 (println "Successful sign up"))
               (println (str student " not able to sign up class : " class-name))))))))

(defn drop-class
  [db student class-name]
  (let [record       ["attends" student class-name]
        seats        (current-class-attendance db class-name)
        class-record ["class" class-name]]
    (tr! db
         (if (get-val tr record)
           (do
             (clear-key tr record)
             (set-val tr class-record (inc (int seats)))
             (println "Successful cancellation"))
           (println (str student " not enrolled in class : " class-name))))))

(defn switch-class
  [db student old-class new-class]
  (tr! db
       (drop-class tr student old-class)
       (signup-class tr student new-class)))

(defn available-classes
  [db]
  (let [record "class"]
    (tr! db
         (filter #(pos-int? (second %1)) (get-range tr record)))))

(comment
  (let [fd (select-api-version 520)]
    (with-open [db (open fd)]
      (tr! db
           (init-db db)
           (add-class db all-classes))))

  ;; signup class
  (let [fd (select-api-version 520)]
    (with-open [db (open fd)]
      (signup-class db "ram" "intro chem 10:00")))

  ;; drop class
  (let [fd (select-api-version 520)]
    (with-open [db (open fd)]
      (drop-class db "ram" "intro chem 10:00")))

  ;; switch class
  (let [fd (select-api-version 520)]
    (with-open [db (open fd)]
      (signup-class db "ram" "intro chem 10:00")
      (switch-class db "ram" "intro chem 10:00" "intro chem 11:00"))))
